File: snippets/cpp/graph/dfs.cpp
std::vector< vertex > dfs( graph const& _graph, vertex const _origin )
{
    std::vector< vertex > res;
    if( res.size() <= static_cast< std::vector< vertex >::size_type >( _origin ) )
    {
        return res;
    }

    std::vector< bool > visited( _graph.size(), false );
    std::stack< vertex > stack;

    stack.push( _origin );
    while( !stack.empty() )
    {
        vertex const v = stack.top();
        stack.pop();

        if( visited[v] )
        {
            continue;
        }
        visited[v] = true;

        res.emplace_back( v );
        std::for_each( std::begin( _graph[v] ),
                       std::end( _graph[v] ),
                       [&stack]( graph::value_type::value_type const& _kv ) { stack.push( _kv.first ); });
    }

    return res;
}

File: contest/snippets/python/graph/dfs.py
def dfs(g, src):
    visited = [False] * len(g)
    stack = [(src, 0)]
    while stack:
        n, c = stack.pop()
        if visited[n]:
            continue
        visited[n] = True

        for _, d, _ in g.edges(n):
            #yield d
            stack.append((d, c + 1))
        yield n, c
